core/src/main/java/kafka/metrics/ClientMetricsInstance.java

```diff
@@ -0,0 +1,122 @@
+/*
+ * Licensed to the Apache Software Foundation (ASF) under one or more
+ * contributor license agreements. See the NOTICE file distributed with
+ * this work for additional information regarding copyright ownership.
+ * The ASF licenses this file to You under the Apache License, Version 2.0
+ * (the "License"); you may not use this file except in compliance with
+ * the License. You may obtain a copy of the License at
+ *
+ *    http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+package kafka.metrics;
+
+import org.apache.kafka.common.Uuid;
+import org.apache.kafka.common.protocol.Errors;
+
+import java.util.Objects;
+import java.util.Set;
+
+/**
+ * Contains the metrics instance metadata and the state of the client instance.
+ */
+public class ClientMetricsInstance {
+
+    private final Uuid clientInstanceId;
+    private final ClientMetricsInstanceMetadata instanceMetadata;
+    private final int subscriptionId;
+    private final long subscriptionUpdateEpoch;
+    private final Set<String> metrics;
+    private final int pushIntervalMs;
+
+    private long lastGetRequestEpoch;
+    private long lastPushRequestEpoch;
+    private volatile boolean terminating;
+    private volatile Errors lastKnownError;
+
+    public ClientMetricsInstance(Uuid clientInstanceId, ClientMetricsInstanceMetadata instanceMetadata,
+        int subscriptionId, long subscriptionUpdateEpoch, Set<String> metrics, int pushIntervalMs) {
+        this.clientInstanceId = Objects.requireNonNull(clientInstanceId);
+        this.instanceMetadata = Objects.requireNonNull(instanceMetadata);
+        this.subscriptionId = subscriptionId;
+        this.subscriptionUpdateEpoch = subscriptionUpdateEpoch;
+        this.metrics = metrics;
+        this.terminating = false;
+        this.pushIntervalMs = pushIntervalMs;
+        this.lastKnownError = Errors.NONE;
+    }
+
+    public Uuid clientInstanceId() {
+        return clientInstanceId;
+    }
+
+    public ClientMetricsInstanceMetadata instanceMetadata() {
+        return instanceMetadata;
+    }
+
+    public int pushIntervalMs() {
+        return pushIntervalMs;
+    }
+
+    public long subscriptionUpdateEpoch() {
+        return subscriptionUpdateEpoch;
+    }
+
+    public int subscriptionId() {
+        return subscriptionId;
+    }
+
+    public Set<String> metrics() {
+        return metrics;
+    }
+
+    public boolean terminating() {
+        return terminating;
```

Comment:
After the instance is terminated, when do we remove the instance form the in-memory state?

Reply:
The reason why I didn't remove the instance from in-memory cache as the subsequent requests from terminated client should be rejected. Having said that, the terminated client should also not remain in-memory forever hence the terminated client should be removed from cache as per the eviction policy of cache i.e. `MAX(60*1000, PushIntervalMs * 3) milliseconds`. This shall be inherently handled by the cache improvement task where we have some time based eviction policy.